language/move-lang/stdlib/modules/libra_system.move

```diff
@@ -0,0 +1,650 @@
+        // check if each validator has rotated their keys, copy their new info and note the change
+        // if so.
+        loop {
+            if (copy_validator_info(validator_info_ref)) {
+                validator_set_changed = true;
+            };
+
+            i = i + 1;
+            if (i >= len) break;
+            validator_info_ref = Vector::borrow_mut(validators_vec_ref, i);
+        };
+
+        // (4) Check for changes in the discovery set
+        i = 0;
+        len = Vector::length(discovery_vec_ref);
+        // assume(len > 0), since an empty validator set is nonsensical
+        let discovery_info_ref = Vector::borrow_mut(discovery_vec_ref, i);
+        // check if each validator has changed their discovery info, copy their new info, and
+        // note the change if so.
+        loop {
+            if (copy_discovery_info(discovery_info_ref)) {
+                discovery_set_changed = true;
+            };
+
+            i = i + 1;
+            if (i >= len) break;
+            discovery_info_ref = Vector::borrow_mut(discovery_vec_ref, i);
+        };
+
+        // (5) Emit validator set changed event if appropriate
+        if (validator_set_changed) {
+            LibraAccount::emit_event<ValidatorSetChangeEvent>(
+                &mut validator_set_ref.change_events,
+                ValidatorSetChangeEvent {
+                    new_validator_set: *validators_vec_ref,
+                },
+            );
+        };
+
+        // (6) Emit discovery set changed event if appropriate
+        if (discovery_set_changed) {
+            LibraAccount::emit_event<DiscoverySetChangeEvent>(
+                &mut discovery_set_ref.change_events,
+                DiscoverySetChangeEvent {
+                    new_discovery_set: *discovery_vec_ref,
+                },
+            );
+        };
+    }
+
+    // Get the ValidatorInfo for the ith validator
+    public fun get_ith_validator_info(i: u64): ValidatorInfo acquires ValidatorSet {
+      let validators_vec_ref = &borrow_global<ValidatorSet>(0x1D8).validators;
+      Transaction::assert(i < Vector::length(validators_vec_ref), 3);
+      *Vector::borrow(validators_vec_ref, i)
+    }
+
+    // Get the address of the i'th validator.
+    public fun get_ith_validator_address(i: u64): address acquires ValidatorSet {
+      let validator_set = borrow_global<ValidatorSet>(0x1D8);
+      let len = Vector::length(&validator_set.validators);
+      Transaction::assert(i < len, 3);
+      Vector::borrow(&validator_set.validators, i).addr
+    }
+
+    // Get the DiscoveryInfo for the ith validator
+    public fun get_ith_discovery_info(i: u64): DiscoveryInfo acquires DiscoverySet {
+        let discovery_vec_ref = &borrow_global<DiscoverySet>(0xD15C0).discovery_set;
+        Transaction::assert(i < Vector::length(discovery_vec_ref), 4);
+        *Vector::borrow(discovery_vec_ref, i)
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
```

Comment:
Since we are porting the modules all together do you think we should move transaction fee to its own module?

Reply:
I know we needed to include it in the system module before, but if we can separate things out that would be great!

However, I would be against doing this at the moment--I don't want to be introducing differences between this code and the IR modules until genesis generation has been swapped over. That way we should be able to generate with either of them. Once the switch is made splitting these makes sense if it's possible.